nucleus/common/common-util/src/main/java/org/glassfish/common/util/InstanceCounter.java

```diff
@@ -0,0 +1,145 @@
+ * General Public License Version 2 only ("GPL") or the Common Development
+ * and Distribution License("CDDL") (collectively, the "License").  You
+ * may not use this file except in compliance with the License.  You can
+ * obtain a copy of the License at
+ * https://github.com/payara/Payara/blob/master/LICENSE.txt
+ * See the License for the specific
+ * language governing permissions and limitations under the License.
+ *
+ * When distributing the software, include this License Header Notice in each
+ * file and include the License file at glassfish/legal/LICENSE.txt.
+ *
+ * GPL Classpath Exception:
+ * The Payara Foundation designates this particular file as subject to the "Classpath"
+ * exception as provided by the Payara Foundation in the GPL Version 2 section of the License
+ * file that accompanied this code.
+ *
+ * Modifications:
+ * If applicable, add the following below the License Header, with the fields
+ * enclosed by brackets [] replaced by your own identifying information:
+ * "Portions Copyright [year] [name of copyright owner]"
+ *
+ * Contributor(s):
+ * If you wish your version of this file to be governed by only the CDDL or
+ * only the GPL Version 2, indicate your decision by adding "[Contributor]
+ * elects to include this software in this distribution under the [CDDL or GPL
+ * Version 2] license."  If you don't indicate a single choice of license, a
+ * recipient has the option to distribute your version of this file under
+ * either the CDDL, the GPL Version 2 or to extend the choice of license to
+ * its licensees as provided above.  However, if you add GPL Version 2 code
+ * and therefore, elected the GPL Version 2 license, then the option applies
+ * only if the new code is made subject to such option by the copyright
+ * holder.
+ */
+package org.glassfish.common.util;
+
+import java.lang.ref.Reference;
+import java.lang.ref.ReferenceQueue;
+import java.lang.ref.WeakReference;
+import java.util.Collections;
+import java.util.IdentityHashMap;
+import java.util.Map;
+import java.util.Set;
+import java.util.concurrent.ConcurrentHashMap;
+
+/**
+ * Counts instances of a particular class, helpful in leak detection
+ * <p>
+ * Example: Put the following line into the class you want counted:
+ * <p>
+ * {@code
+ *     private final InstanceCounter<MyClass> instanceCounter = new InstanceCounter<>(this);
+ * }
+ *
+ * @author Cuba Stanley
+ */
+public class InstanceCounter {
+    private static class Count {
+        int instanceCount;
+        private final ReferenceQueue<Object> referenceQueue = new ReferenceQueue<>();
+        private final Set<WeakReference<?>> references = Collections.newSetFromMap(new IdentityHashMap<>());
+
+        public Count() {
+            instanceCount = 1;
+        }
+
+        public Count(int instanceCount) {
+            this.instanceCount = instanceCount;
+        }
+    }
+
+    private static final Map<Class<?>, Count> INSTANCE_COUNT = new ConcurrentHashMap<>();
```

Comment:
I would use ConcurrentMap interface to point that there is expected concurrency.

```suggestion
    private static final ConcurrentMap<Class<?>, Count> INSTANCE_COUNT = new ConcurrentHashMap<>();
```